storage/postgres/notificator.go

```diff
@@ -103,73 +109,133 @@ func (n *Notificator) Start(ctx context.Context, group *sync.WaitGroup) error {
+	}
+	queue, err := storage.NewNotificationQueue(n.queueSize)
 	if err != nil {
-		return nil, invalidRevisionNumber, fmt.Errorf("could not get platform from user context %v", err)
+		return nil, types.INVALIDREVISION, err
 	}
-	if platform.ID == "" {
-		return nil, invalidRevisionNumber, errors.New("platform ID not found in user context")
+	if err = n.startListening(); err != nil {
+		return nil, types.INVALIDREVISION, fmt.Errorf("listen to %s channel failed %v", postgresChannel, err)
 	}
-	if atomic.LoadInt32(&n.isConnected) == aFalse {
-		return nil, invalidRevisionNumber, errors.New("cannot register consumer - Notificator is not running")
+	lastKnownRevisionToSM := n.addConsumer(platform, queue)
+	if lastKnownRevision == types.INVALIDREVISION {
+		return queue, lastKnownRevisionToSM, nil
 	}
-	if err = n.startListening(); err != nil {
-		return nil, invalidRevisionNumber, fmt.Errorf("listen to %s channel failed %v", postgresChannel, err)
+	queueWithMissedNotifications, err := n.replaceQueueWithMissingNotificationsQueue(queue, lastKnownRevision, lastKnownRevisionToSM, platform)
+	if err != nil {
+		if errUnregisterConsumer := n.UnregisterConsumer(queue); errUnregisterConsumer != nil {
+			log.C(n.ctx).WithError(errUnregisterConsumer).Errorf("Could not unregister notification consumer %s", queue.ID())
+		}
+		return nil, types.INVALIDREVISION, err
 	}
-	queue, err := storage.NewNotificationQueue(n.queueSize)
+	return queueWithMissedNotifications, lastKnownRevisionToSM, nil
+}
+
+func (n *Notificator) filterRecipients(recipients []*types.Platform, notification *types.Notification) []*types.Platform {
+	for _, filter := range n.notificationFilters {
+		recipients = filter(recipients, notification)
+		if len(recipients) == 0 {
+			return recipients
+		}
+	}
+	return recipients
+}
+
+func (n *Notificator) replaceQueueWithMissingNotificationsQueue(queue storage.NotificationQueue, lastKnownRevision, lastKnownRevisionToSM int64, platform *types.Platform) (storage.NotificationQueue, error) {
+	if lastKnownRevision > lastKnownRevisionToSM {
+		log.C(n.ctx).Debug("lastKnownRevision is grater than the one SM knows")
+		return nil, util.ErrInvalidNotificationRevision
+	}
+	if _, err := n.storage.GetNotificationByRevision(n.ctx, lastKnownRevision); err != nil {
+		if err == util.ErrNotFoundInStorage {
+			log.C(n.ctx).WithError(err).Debugf("notification with revision %d not found in storage", lastKnownRevision)
+			return nil, util.ErrInvalidNotificationRevision
+		}
+		return nil, err
+	}
+
+	missedNotifications, err := n.storage.ListNotifications(n.ctx, platform.ID, lastKnownRevision, lastKnownRevisionToSM)
 	if err != nil {
-		return nil, invalidRevisionNumber, err
+		return nil, err
+	}
+	filteredMissedNotification := make([]*types.Notification, 0, len(missedNotifications))
+	var recipients []*types.Platform
+	for _, notification := range missedNotifications {
+		recipients = n.filterRecipients([]*types.Platform{platform}, notification)
```

Comment:
this is very weird... it says "filter notifications" but filters platforms and after that this filtering doesn't use the returned result. The problem is that you're using the same function to do 2 different things.

Reply:
Hopefully with the rename of `NotificationFilterFunc` to `ReceiversFilterFunc` it might be more clear. The idea is to filter given recipients for a specific notification by returning the filtered recipients. This way the one who is writing the filter can process the notification once for all recipients. `filterRecipients` function iterates over all registered filter functions and applies them to the initial recipients and given notification.